flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+	 */
+	public void remove(String pathInZooKeeper, BackgroundCallback callback) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		checkNotNull(callback, "Background callback");
+
+		client
+				.delete()
+				.deletingChildrenIfNeeded()
+				.inBackground(callback)
+				.forPath(pathInZooKeeper);
+	}
+
+	/**
+	 * Discards a state handle and removes it from ZooKeeper.
+	 *
+	 * <p><strong>Important</strong>: This method will only discard the state handle and not the
+	 * state itself. Don't forget to run custom cleanup code of the state, if necessary.
+	 *
+	 * <pre>
+	 * T state = get(path).getState();
+	 * state.discard(); // Custom clean up
+	 * discard(path)
+	 * </pre>
+	 *
+	 * <p>If you only want to remove the state handle in ZooKeeper call {@link #remove(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to discard (expected to start with a '/')
+	 * @throws Exception
+	 */
+	public void discard(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		StateHandle<T> stateHandle = get(pathInZooKeeper);
+
+		// Delete the state handle from ZooKeeper first
+		client
+				.delete()
+				.deletingChildrenIfNeeded()
+				.forPath(pathInZooKeeper);
+
+		// Discard the state handle only after it has been successfully deleted from ZooKeeper.
+		// Otherwise we might enter an illegal state after failures (with a state handle in
+		// ZooKeeper, which has already been discarded).
+		stateHandle.discardState();
+	}
+
+	/**
+	 * Discards all available state handles and removes them from ZooKeeper.
+	 *
+	 * <p><strong>Important</strong>: This method will only discard the state handles and not the
+	 * states itself. Don't forget to run custom cleanup code of the states, if necessary.
+	 *
+	 * <pre>
+	 * for (StateHandle<T> handle : getAll()) {
+	 *     handle.getState().discard(); // Custom clean up
+	 * }
+	 * discardAll()
+	 * </pre>
+	 *
+	 * @throws Exception
+	 */
+	public void discardAll() throws Exception {
+		final List<ZooKeeperStateHandle<T>> allStateHandles = getAll();
+
+		ZKPaths.deleteChildren(
+				client.getZookeeperClient().getZooKeeper(),
+				ZKPaths.fixForNamespace(client.getNamespace(), "/"),
+				false);
+
+		// Discard the state handles only after they have been successfully deleted from ZooKeeper.
+		for (ZooKeeperStateHandle<T> stateHandle : allStateHandles) {
+			stateHandle.getStateHandle().discardState();
+		}
+	}
+
+	/**
+	 * Throws a {@link ConcurrentModificationException} if the cVersion of the specified path
+	 * does not match the expected cVersion.
+	 *
+	 * @param expectedCVersion The expected cVersion
+	 * @param pathInZooKeeper  Path in ZooKeeper to check (expected to exist and start with '/')
+	 * @param errorMessage     The error message of the thrown Exception
+	 * @throws Exception
+	 */
+	private void verifyExpectedCVersion(
+			int expectedCVersion,
+			String pathInZooKeeper,
+			String errorMessage) throws Exception {
```

Comment:
Most of the generic Exception comes from Curator. I have to check whether it is feasible to provide more specific ones.